### PR TITLE
overleaf: init at 4.2 & nixos/overleaf: init

### LINKED ISSUE #207329: Packaging request: overleaf
**Project description**
Open-source and self-hostable online real-time collaborative LaTeX editor.

It would be great to have it packaged for NixOS as well as a NixOS module to quickly configure it.

**Metadata**

* homepage URL: https://www.overleaf.com/
* source URL: https://github.com/overleaf/overleaf
* license: agpl3
* platforms: linux

### PR DESCRIPTION
**Pull request by JulienMalka**
###### Description of changes

This PR adds the overleaf package and module (Fixes #207329).
This work is done in collaboration with @camillemndn.

NPM was unable to generate a correct ``package-lock.json`` file for the project so we had to convert the project to yarn.

TODO: 
- [ ] Fix Darwin build
- [ ] Write the module
- [ ] Write a nixos test
- [ ] Write release notes

Thanks to @winterqt, @lilyinstarlight and @mweinelt for their help during times of darkness.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/overleaf.nix [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        "filestore"
+        "notifications"
+        "project-history"
+        "real-time"
+        "spelling"
+        "track-changes"
+        "web"
+      ]) ++
+      [{
+        ferretdb = mkIf (cfg.mongoDB == "ferretdb") {
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-apps/overleaf.nix:134]: Early nitpick: I would probably split up the FerretDB specific stuff into another module in the following way:
```
config = mkMerge [{ general configuration ... } (mkIf cfg.mongoDB == "ferretdb" { FerretDB specific customizations... }]
```

this way, you would have less if-guards

Comment (JulienMalka) [reply nixos/modules/services/web-apps/overleaf.nix:134]: I'm doing a ferretDB module as we speak

File: pkgs/servers/overleaf/default.nix [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    rm package-lock.json
+    cp ${./yarn.lock} yarn.lock
+    chmod +rw yarn.lock
+    cp ${./package.json} package.json
+
+    # Replace hard-coded values in settings by environment variables
+    find . -type f -not -path '*/\.*' -exec sed -E -i "s|SHARELATEX_|OVERLEAF_|g" {} +
+    (
+      cd server-ce/config
+      sed -E -i "s!'dockerhost',!undefined,\n      path: process.env.OVERLEAF_REDIS_PATH || undefined,!" ./settings.js
```
Comment (RaitoBezarius) [inline pkgs/servers/overleaf/default.nix:37]: How feasible do you think we can send a upstream patch for those?

File: pkgs/servers/overleaf/default.nix [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      --add-flags app.js \
+      --chdir $out/share/services/web
+  '';
+
+  meta = with lib; {
+    description = "A web-based collaborative LaTeX editor";
+    homepage = "https://github.com/overleaf/overleaf";
+    license = licenses.agpl3Only;
+    maintainers = with maintainers; [ camillemndn julienmalka ];
+    platforms = platforms.all;
```
Comment (RaitoBezarius) [inline pkgs/servers/overleaf/default.nix:142]: it's lower-bound by nodejs I think :P

File: nixos/modules/services/web-apps/overleaf.nix [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    };
+
+    mongoDB = mkOption {
+      type = types.str;
+      default = "mongodb";
+      example = "ferretdb";
+      description = mdDoc ''The type of MongoDB to enable. One of "mongodb", "ferretdb" or "none"'';
+    };
+
+    texliveScheme = mkOption {
```
Comment (RaitoBezarius) [inline nixos/modules/services/web-apps/overleaf.nix:27]: could we not just pass the proper texlive packages we exactly want? sounds easier than mandating using a certain path

File: nixos/modules/services/web-apps/overleaf.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nixos/modules/services/web-apps/overleaf.nix:134]: Amazing :) (but my comment still hold regarding minimizing the if-guards to make it easy to review FerretDB-specific customizations)

File: nixos/modules/services/web-apps/overleaf.nix (hunk 4 above)
Comment (camillemndn) [reply nixos/modules/services/web-apps/overleaf.nix]: Done! Julien actually scolded me for this too :p

File: pkgs/servers/overleaf/default.nix (hunk 2 above)
Comment (JulienMalka) [reply pkgs/servers/overleaf/default.nix:37]: Probably not feasible, we are kinda tweaking the files they use to create their docker image: we replace the ``SHARELATEX_*`` env variables by ``OVERLEAF_*``, this is just cosmetic not to confuse the module users.
We also tweak the ``settings.js`` for more customizations. The overleaf ``settings.js`` file assumes running into docker and doesn't expose all options (for example when not running in a docker, it is nice to be able to change to web interface port, the port 3000 could be used by something else). The other alternative to using this script would be to commit our own ``settings.js`` to nixpkgs, which would be cleaner in my opinion but would have to merge conflicts at each update.

Comment (JulienMalka) [reply pkgs/servers/overleaf/default.nix:37]: I've submitted a patch upstream to avoid ``versions.patch``, we'll see if they are interested.

**Review by camillemndn — COMMENTED** [review]
The package and module are nearly ready, but maybe it should be made more explicit that the Community Edition does not easily allow restricting access to their project compilation folder on the disk, meaning each Overleaf user has basically the same system permissions as the systemd unit user.

One way to prevent this is to use Dockerized images of TeXLive, which I have tested but we cannot require NixOS users to enable Docker, and it also slows the compilation process.

Maybe there are ways to use systemd for this, but it seems to require patching the source code quite a lot.

**Comment by RaitoBezarius** [comment]
@mweinelt What do you think about adding a `knownVulnerabilities` for this network service? Technically, it is a serious vuln IMHO. I don't know how to convey this information to potential users.

**Comment by RaitoBezarius** [comment]
@camillemndn You still have some ofborg evaluation errors :).

**Comment by mweinelt** [comment]
@RaitoBezarius I appear to be lacking context.

**Comment by camillemndn** [comment]
> @camillemndn You still have some ofborg evaluation errors :).

Woops sorry I just realized that the ```nodejs-XX_x``` packages have been renamed!

**Comment by RaitoBezarius** [comment]
> @RaitoBezarius I appear to be lacking context.

> but maybe it should be made more explicit that the Community Edition does not easily allow restricting access to their project compilation folder on the disk, meaning each Overleaf user has basically the same system permissions as the systemd unit user.

Overleaf in CE enables each user to read anything in the scope of the systemd unit, e.g. any user could potentially read another user's data.

File: pkgs/servers/overleaf/default.nix [hunk 5]
```diff
@@ -15,6 +15,18 @@ let
...
   version = "3.5";
 
+  inherit (stdenv.hostPlatform) system;
+  throwSystem = throw "Unsupported system: ${system}";
+
+  sha256 = {
+    x86_64-linux = "sha256-ci68170df9o3fJuw8AH/qdJR8NZ2o9OeAg8gfjycHd0=
+";
+    aarch64-linux = "sha256-ci68170df9o3fJuw8AH/qdJR8NZ2o9OeAg8gfjycHd0=
+";
```
Comment (mweinelt) [inline pkgs/servers/overleaf/default.nix:25]:
```suggestion
    x86_64-linux = "sha256-ci68170df9o3fJuw8AH/qdJR8NZ2o9OeAg8gfjycHd0=";
    aarch64-linux = "sha256-ci68170df9o3fJuw8AH/qdJR8NZ2o9OeAg8gfjycHd0=";
```

**Comment by mweinelt** [comment]
> > @RaitoBezarius I appear to be lacking context.
> 
> > but maybe it should be made more explicit that the Community Edition does not easily allow restricting access to their project compilation folder on the disk, meaning each Overleaf user has basically the same system permissions as the systemd unit user.
> 
> Overleaf in CE enables each user to read anything in the scope of the systemd unit, e.g. any user could potentially read another user's data.

Not really a vulnerability, it just makes it a public instance and everyone should be made aware what that means.

**Comment by pbsds** [comment]
> Not really a vulnerability, it just makes it a public instance and everyone should be made aware what that means.

Perhaps mention it in the nixos module `enable` option, and encourage the user to consider running it in a nixos container.

**Comment by JulienMalka** [comment]
Thanks for these feedbacks :) We will cleanup that PR and make it ready for review :)

**Comment by felbinger** [comment]
Any news on this?

**Comment by bhankas** [comment]
any update on state of things? It's kinda boring to use docker just for overleaf..

**Comment by camillemndn** [comment]
I am closing this, as I do not plan to finish it soon

**Comment by pbsds** [comment]
Sad to see this close, but thank you for all your efforts. :)

**Comment by raks0009** [comment]
> any update on state of things? It's kinda boring to use docker just for overleaf..

a nixos noob here :( but would you be willing to explain how you were able to set up the overleaf docker @bhankas  ? thanks!
there are many guides, but very less clarity for me.
thanks for your efforts btw @camillemndn and @JulienMalka

**Comment by Yeshey** [comment]
> > any update on state of things? It's kinda boring to use docker just for overleaf..
> 
> a nixos noob here :( but would you be willing to explain how you were able to set up the overleaf docker @bhankas ? thanks! there are many guides, but very less clarity for me. thanks for your efforts btw @camillemndn and @JulienMalka

You can checkout my solution to run overleaf with docker here: https://github.com/Yeshey/nixOS-Config/blob/main/modules/nixos/toHost/overleaf.nix, maybe it can help you
But I'm doing it in a way to work on aarch64, i havent tested in x86, also its pulling a fork i made to have full latex support instead of the upstream overleaf

**Pull request closed without merging**